In [1]:
import pandas as pd
pd.set_option('max_column', None)
import numpy as np
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

In [2]:
url = 'https://raw.githubusercontent.com/katenjoki/Telecom_Week1Challenge/main/OneDrive/Desktop/10Academy/Telecom_Week1Challenge/data/Week1_challenge_data_source(CSV).csv'
data = pd.read_csv(url)
data.head()

Bearer Id            Start  Start ms              End  End ms  \
0  1.311448e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   662.0   
1  1.311448e+19   4/9/2019 13:04     235.0   4/25/2019 8:15   606.0   
2  1.311448e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   652.0   
3  1.311448e+19   4/10/2019 0:31     486.0   4/25/2019 7:36   171.0   
4  1.311448e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   954.0   

   Dur. (ms)          IMSI  MSISDN/Number          IMEI  \
0  1823652.0  2.082014e+14   3.366496e+10  3.552121e+13   
1  1365104.0  2.082019e+14   3.368185e+10  3.579401e+13   
2  1361762.0  2.082003e+14   3.376063e+10  3.528151e+13   
3  1321509.0  2.082014e+14   3.375034e+10  3.535661e+13   
4  1089009.0  2.082014e+14   3.369980e+10  3.540701e+13   

      Last Location Name  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0  9.16456699548519E+015             42.0              5.0   
1                L77566A             65.0              5.0   
2                D42335A              NaN              NaN   
3                T21824A              NaN              NaN   
4                D88865A              NaN              NaN   

   Avg Bearer TP DL (kbps)  Avg Bearer TP UL (kbps)  \
0                     23.0                     44.0   
1                     16.0                     26.0   
2                      6.0                      9.0   
3                     44.0                     44.0   
4                      6.0                      9.0   

   TCP DL Retrans. Vol (Bytes)  TCP UL Retrans. Vol (Bytes)  \
0                          NaN                          NaN   
1                          NaN                          NaN   
2                          NaN                          NaN   
3                          NaN                          NaN   
4                          NaN                          NaN   

   DL TP < 50 Kbps (%)  50 Kbps < DL TP < 250 Kbps (%)  \
0                100.0                             0.0   
1                100.0                             0.0   
2                100.0                             0.0   
3                100.0                             0.0   
4                100.0                             0.0   

   250 Kbps < DL TP < 1 Mbps (%)  DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  \
0                            0.0                 0.0                100.0   
1                            0.0                 0.0                100.0   
2                            0.0                 0.0                100.0   
3                            0.0                 0.0                100.0   
4                            0.0                 0.0                100.0   

   10 Kbps < UL TP < 50 Kbps (%)  50 Kbps < UL TP < 300 Kbps (%)  \
0                            0.0                             0.0   
1                            0.0                             0.0   
2                            0.0                             0.0   
3                            0.0                             0.0   
4                            0.0                             0.0   

   UL TP > 300 Kbps (%)  HTTP DL (Bytes)  HTTP UL (Bytes)  \
0                   0.0              NaN              NaN   
1                   0.0              NaN              NaN   
2                   0.0              NaN              NaN   
3                   0.0              NaN              NaN   
4                   0.0              NaN              NaN   

   Activity Duration DL (ms)  Activity Duration UL (ms)   Dur. (ms).1  \
0                    37624.0                    38787.0  1.823653e+09   
1                      168.0                     3560.0  1.365104e+09   
2                        0.0                        0.0  1.361763e+09   
3                     3330.0                    37882.0  1.321510e+09   
4                        0.0                        0.0  1.089009e+09   

  Handset Manufacturer                 Handset Type  \
0              Samsung   Samsung Galaxy A5 Sm-A520F   
1   

In [3]:
data.shape

(150001, 55)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

<h2>Task 1 : User Overview Analysis </h2>

<h3> Data Cleaning </h3>

In [5]:
data.columns = data.columns.str.replace(' ','_')
data.columns.to_list()

['Bearer_Id',
 'Start',
 'Start_ms',
 'End',
 'End_ms',
 'Dur._(ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last_Location_Name',
 'Avg_RTT_DL_(ms)',
 'Avg_RTT_UL_(ms)',
 'Avg_Bearer_TP_DL_(kbps)',
 'Avg_Bearer_TP_UL_(kbps)',
 'TCP_DL_Retrans._Vol_(Bytes)',
 'TCP_UL_Retrans._Vol_(Bytes)',
 'DL_TP_<_50_Kbps_(%)',
 '50_Kbps_<_DL_TP_<_250_Kbps_(%)',
 '250_Kbps_<_DL_TP_<_1_Mbps_(%)',
 'DL_TP_>_1_Mbps_(%)',
 'UL_TP_<_10_Kbps_(%)',
 '10_Kbps_<_UL_TP_<_50_Kbps_(%)',
 '50_Kbps_<_UL_TP_<_300_Kbps_(%)',
 'UL_TP_>_300_Kbps_(%)',
 'HTTP_DL_(Bytes)',
 'HTTP_UL_(Bytes)',
 'Activity_Duration_DL_(ms)',
 'Activity_Duration_UL_(ms)',
 'Dur._(ms).1',
 'Handset_Manufacturer',
 'Handset_Type',
 'Nb_of_sec_with_125000B_<_Vol_DL',
 'Nb_of_sec_with_1250B_<_Vol_UL_<_6250B',
 'Nb_of_sec_with_31250B_<_Vol_DL_<_125000B',
 'Nb_of_sec_with_37500B_<_Vol_UL',
 'Nb_of_sec_with_6250B_<_Vol_DL_<_31250B',
 'Nb_of_sec_with_6250B_<_Vol_UL_<_37500B',
 'Nb_of_sec_with_Vol_DL_<_6250B',
 'Nb_of_sec_with_Vol_UL_<_1250B',
 'Socia

**Checking for null values**

In [6]:
#Check if there are rows where all column values are null
check_null = data.columns.isnull().any()
check_null

False

In [7]:
#% of null values
def null_analysis(df):
    total_values = np.product(df.shape)
    total_null = (df.isnull().sum()).sum()
    proportion_null = (total_null / total_values)*100
    
    print("The Telcommunication dataset has ",proportion_null,"% null values.")

null_analysis(data)

The Telcommunication dataset has  12.501637867868759 % null values.


**Finding columns that have more than 30% of null values**

In [8]:
def null_per_column(df):
    dataset = df.isnull().sum()
    dataset = pd.DataFrame(dataset,columns=['null'])
    dataset['%null'] = (dataset['null']/df.shape[0])*100
    dataframe = dataset[dataset['%null']>=30]
    print(dataframe)    
    #print('\n List of columns with more than 30% null values:\n',dataframe.index.to_list())

null_per_column(data)

                                            null      %null
TCP_DL_Retrans._Vol_(Bytes)                88146  58.763608
TCP_UL_Retrans._Vol_(Bytes)                96649  64.432237
HTTP_DL_(Bytes)                            81474  54.315638
HTTP_UL_(Bytes)                            81810  54.539636
Nb_of_sec_with_125000B_<_Vol_DL            97538  65.024900
Nb_of_sec_with_1250B_<_Vol_UL_<_6250B      92894  61.928920
Nb_of_sec_with_31250B_<_Vol_DL_<_125000B   93586  62.390251
Nb_of_sec_with_37500B_<_Vol_UL            130254  86.835421
Nb_of_sec_with_6250B_<_Vol_DL_<_31250B     88317  58.877607
Nb_of_sec_with_6250B_<_Vol_UL_<_37500B    111843  74.561503


In [9]:
#Dropping columns with more than 30% null values
data=data.drop(['TCP_DL_Retrans._Vol_(Bytes)', 'TCP_UL_Retrans._Vol_(Bytes)', 'HTTP_DL_(Bytes)', 'HTTP_UL_(Bytes)', 'Nb_of_sec_with_125000B_<_Vol_DL', 'Nb_of_sec_with_1250B_<_Vol_UL_<_6250B', 'Nb_of_sec_with_31250B_<_Vol_DL_<_125000B', 'Nb_of_sec_with_37500B_<_Vol_UL', 'Nb_of_sec_with_6250B_<_Vol_DL_<_31250B', 'Nb_of_sec_with_6250B_<_Vol_UL_<_37500B'],axis=1)
data.shape

(150001, 45)

In [10]:
data.isnull().sum()

Bearer_Id                           991
Start                                 1
Start_ms                              1
End                                   1
End_ms                                1
Dur._(ms)                             1
IMSI                                570
MSISDN/Number                      1066
IMEI                                572
Last_Location_Name                 1153
Avg_RTT_DL_(ms)                   27829
Avg_RTT_UL_(ms)                   27812
Avg_Bearer_TP_DL_(kbps)               1
Avg_Bearer_TP_UL_(kbps)               1
DL_TP_<_50_Kbps_(%)                 754
50_Kbps_<_DL_TP_<_250_Kbps_(%)      754
250_Kbps_<_DL_TP_<_1_Mbps_(%)       754
DL_TP_>_1_Mbps_(%)                  754
UL_TP_<_10_Kbps_(%)                 792
10_Kbps_<_UL_TP_<_50_Kbps_(%)       792
50_Kbps_<_UL_TP_<_300_Kbps_(%)      792
UL_TP_>_300_Kbps_(%)                792
Activity_Duration_DL_(ms)             1
Activity_Duration_UL_(ms)             1
Dur._(ms).1                           1


* Column Bearer Id has 991 values, we drop these rows as the Bearer Id should be a unique identifier

In [11]:
data = data[data['Bearer_Id'].notna()]
data.isnull().sum()

Bearer_Id                             0
Start                                 0
Start_ms                              0
End                                   0
End_ms                                0
Dur._(ms)                             0
IMSI                                  9
MSISDN/Number                       504
IMEI                                 11
Last_Location_Name                  162
Avg_RTT_DL_(ms)                   27685
Avg_RTT_UL_(ms)                   27666
Avg_Bearer_TP_DL_(kbps)               0
Avg_Bearer_TP_UL_(kbps)               0
DL_TP_<_50_Kbps_(%)                 716
50_Kbps_<_DL_TP_<_250_Kbps_(%)      716
250_Kbps_<_DL_TP_<_1_Mbps_(%)       716
DL_TP_>_1_Mbps_(%)                  716
UL_TP_<_10_Kbps_(%)                 770
10_Kbps_<_UL_TP_<_50_Kbps_(%)       770
50_Kbps_<_UL_TP_<_300_Kbps_(%)      770
UL_TP_>_300_Kbps_(%)                770
Activity_Duration_DL_(ms)             0
Activity_Duration_UL_(ms)             0
Dur._(ms).1                           0


In [12]:
#replace null values for object types with mode
def null_objects(df,column):
    df[column] = df[column].fillna(df[column].mode()[0])
    return df[column]

data['Last_Location_Name'] = null_objects(data,'Last_Location_Name')
data['Handset_Manufacturer'] = null_objects(data,'Handset_Manufacturer')
data['Handset_Type']= null_objects(data,'Handset_Type')

In [13]:
#list of columns with null values,
def null_values(df):
    dataframe=df.isnull().sum()
    dataframe=pd.DataFrame(dataframe,columns=['null'])
    dataframe=dataframe[dataframe['null']>0]
    print(dataframe.index.to_list())

null_values(data)

['IMSI', 'MSISDN/Number', 'IMEI', 'Avg_RTT_DL_(ms)', 'Avg_RTT_UL_(ms)', 'DL_TP_<_50_Kbps_(%)', '50_Kbps_<_DL_TP_<_250_Kbps_(%)', '250_Kbps_<_DL_TP_<_1_Mbps_(%)', 'DL_TP_>_1_Mbps_(%)', 'UL_TP_<_10_Kbps_(%)', '10_Kbps_<_UL_TP_<_50_Kbps_(%)', '50_Kbps_<_UL_TP_<_300_Kbps_(%)', 'UL_TP_>_300_Kbps_(%)', 'Nb_of_sec_with_Vol_DL_<_6250B', 'Nb_of_sec_with_Vol_UL_<_1250B']


In [14]:
#check skewness
data_test=data[['IMSI', 'MSISDN/Number', 'IMEI', 'Avg_RTT_DL_(ms)', 'Avg_RTT_UL_(ms)', 'DL_TP_<_50_Kbps_(%)', '50_Kbps_<_DL_TP_<_250_Kbps_(%)', '250_Kbps_<_DL_TP_<_1_Mbps_(%)', 'DL_TP_>_1_Mbps_(%)', 'UL_TP_<_10_Kbps_(%)', '10_Kbps_<_UL_TP_<_50_Kbps_(%)', '50_Kbps_<_UL_TP_<_300_Kbps_(%)', 'UL_TP_>_300_Kbps_(%)', 'Nb_of_sec_with_Vol_DL_<_6250B', 'Nb_of_sec_with_Vol_UL_<_1250B']]
data_test.skew()

IMSI                               40.987487
MSISDN/Number                     331.677128
IMEI                                1.070670
Avg_RTT_DL_(ms)                    66.756066
Avg_RTT_UL_(ms)                    28.499279
DL_TP_<_50_Kbps_(%)                -2.298563
50_Kbps_<_DL_TP_<_250_Kbps_(%)      3.273358
250_Kbps_<_DL_TP_<_1_Mbps_(%)       4.571437
DL_TP_>_1_Mbps_(%)                  5.361127
UL_TP_<_10_Kbps_(%)                -8.985914
10_Kbps_<_UL_TP_<_50_Kbps_(%)      10.946940
50_Kbps_<_UL_TP_<_300_Kbps_(%)     21.833582
UL_TP_>_300_Kbps_(%)               36.151871
Nb_of_sec_with_Vol_DL_<_6250B       9.024910
Nb_of_sec_with_Vol_UL_<_1250B       7.482728
dtype: float64

In [15]:
#since data is skewed we replace the null values with the median
column_names = ['IMSI', 'MSISDN/Number', 'IMEI', 'Avg_RTT_DL_(ms)', 'Avg_RTT_UL_(ms)', 'DL_TP_<_50_Kbps_(%)', '50_Kbps_<_DL_TP_<_250_Kbps_(%)', '250_Kbps_<_DL_TP_<_1_Mbps_(%)', 'DL_TP_>_1_Mbps_(%)', 'UL_TP_<_10_Kbps_(%)', '10_Kbps_<_UL_TP_<_50_Kbps_(%)', '50_Kbps_<_UL_TP_<_300_Kbps_(%)', 'UL_TP_>_300_Kbps_(%)', 'Nb_of_sec_with_Vol_DL_<_6250B', 'Nb_of_sec_with_Vol_UL_<_1250B']

for i in column_names:
    data[i]=data[i].fillna(data[i].mean())

(data.isnull().sum()).sum()

0